mrouter_host: Add test

wkz opened the pull request:
Generalize the existing mrouter_port test, such that the router port can be either a bridge port, or the bridge itself.

This means we have to deal with all the nasty inconsistencies of allowed flags for MDB entries etc. It also means that while we can use the bridge's internal querier function for the port test, we must generate our own queries when the querier is placed behind a bridge port. For this reason, the host test requires that nemesis is available on the test system.

wkz committed:
mrouter_host: Add test

Generalize the existing mrouter_port test, such that the router port
can be either a bridge port, or the bridge itself.

This means we have to deal with all the nasty inconsistencies of
allowed flags for MDB entries etc. It also means that while we can use
the bridge's internal querier function for the port test, we must
generate our own queries when the querier is placed behind a bridge
port. For this reason, the host test requires that nemesis is
available on the test system.

cappe987 commented on suite/40-mrouter.sh:
```diff
@@ -77,104 +104,138 @@ mrouter_report_one()
+    mode=$1
 
-    local bropts="mcast_snooping 1"
+    ib=$b1
+    ih=$h1
+    ob=$b2
+    oh=$h2
+    qb=$([ $mode = host ] && echo $b3 || echo $br0)
+    qh=$([ $mode = host ] && echo $h3)
+    rb=$([ $mode = port ] && echo $b3 || echo $br0)
+    rh=$([ $mode = port ] && echo $h3 || echo $br0)
```
Perhaps add a comment explaining the meanings of `i`, `o`, `q`, `r`.

wkz replied:
Good point! I'll add a legend with descriptions.

cappe987 commented on suite/40-mrouter.sh:
```diff
@@ -77,104 +104,138 @@ mrouter_report_one()
+    rb=$([ $mode = port ] && echo $b3 || echo $br0)
+    rh=$([ $mode = port ] && echo $h3 || echo $br0)
+
+    local bropts="mcast_snooping 1 mcast_flood_mrouters_only 0"
```
`mcast_flood_mrouters_only` is an internal patch. Maybe add a check if it exists and skip the test if not, or output a warning and create the bridge without it. Now it will just fail to create the bridge if not present. Since it only affects when running in software it would be fine if it allows the test to still run when testing hardware.

wkz replied:
Right, I keep forgetting that. This was part of a series from @troglobit, of which only half was merged upstream IIRC. I'll work out a better fallback.

Thanks!

wkz replied:
I was great that you pointed this out. It got me to dig a bit deeper, and it will most likely result in us dropping this patch and just handle flooding in the way that the upstream bridge does.

The test now expects the behavior corresponding to `mcast_flood_mrouters_only=1`, _but_, it accepts (with a warning) systems that behave as if `mcast_flood_mrouters_only=0` (e.g. offloaded bridges on an mv88e6xxx system).

Fair?

cappe987 replied:
Well done! Showing warnings seems like the best way to solve this issue since it allows proper testing of all the cases.

wkz committed:
lib: Add helpers to check for bridge feature availability

br_has checks for a feature on the bridge master, while brport_has
checks for a feature on bridge ports.

wkz committed:
lib: Add helper to test a current bridge/brport setting

E.g. `br_test $b2 mast_flood on` will return 0 iff multicast flooding
is enabled on $b2. The same helper can also be used on a bridge
interface.
